### ISSUE TITLE
Norm metric calculation incorrect for UCF101 video inputs

### ISSUE THREAD

**Issue by yusong-tan** [twosixlabs/armory#715]
The `norm` function defined in https://github.com/twosixlabs/armory/blob/91f60420cdb7113e4cfddd7127c79e7280aac7e0/armory/utils/metrics.py#L74 assumes argument `x` has batch_size as the first dimension and the rest of the dimensions are associated with the model input (e.g., for image, `x` has shape `(nb, h, w, c)`, and each model input has shape `(h, w, c)`).

However, currently the way the video scenario is structured, whole videos are batched and passed into the norm function, but each video actually represents multiple model inputs. So `x` has shape `(nb, nf, 3, 16, 112, 112)`, and each model input has shape `(3, 16, 112, 112)`. When `norm` is called on `x`, it actually calculates one norm value over multiple inputs. This is only a problem since perturbation norms are based on a single input and not over multiple inputs. In order to correctly log the correct norm calculated by an attack algorithm, the `norm` function needs to treat the first two dimensions as batches.

**Comment by davidslater** [comment]
Can you clarify what you're looking at w.r.t. what the desired output is. If we calculate a separate norm across each frame, we still need to combine them at the end so that we have a single norm associated with a particular datapoint. This could involve taking per-frame norms and then combining them via average or max or some other aggregation function.

For most metrics, I think that a simple average across frames would make sense. However, Linf seems like max would make more sense. Thoughts?

**Comment by davidslater** [comment]
What do the terms `nf` and `16` refer to? I had thought that this generated a variable number of frames, but I don't understand where the 16 comes from.

**Comment by davidslater** [comment]
Ah, I see - they are numbers of stacks of 16-frames.

This makes it more confusing as to how to aggregate the metrics together. Do we take the average over the 16 frames and then take the average again over the stacks?

**Comment by davidslater** [comment]
I have a candidate example that reshapes before using the image metrics. 

https://github.com/twosixlabs/armory/pull/717/files

**Comment by yusong-tan** [comment]
I don't think the reshaping and using the image metrics works for L2 norm. For L2 on video, it is calculated as sqrt(sum( (x_i)^2 )), where i = 0,...,3\*16\*112\*112. This is not the same as mean(sqrt(sum( (x_i_j)^2 ))), where i=0,...,15 and j=0,...,3\*112\*112.

Currently in the video scenario, x is of shape (nb, num_stacks, 3, 16, 112, 112). To calculate norm correctly for each model input, it needs to be reshaped into (nb\*num_stacks, 3, 16, 112, 112). If one wants to calculate an average norm value for each video, then num_stacks consecutive metrics will need to be averaged, but num_stacks is a variable number and changes from video to video, so care needs to be taken.

**Comment by davidslater** [comment]
Okay,  I think I have a solution now. Check out https://github.com/twosixlabs/armory/pull/717/files specifically the metric `"mars_mean_l2"`. Let me know what you think @yusong-tan

**Comment by yusong-tan** [comment]
@davidslater Yes, I think `mars_mean_l2` works. In fact, I think it'll work as long as `x` is of the shape `(n_batch, n_stacks, _, _, _, _)`, so `verify_mars()` may not be necessary as it won't work with models that take different input shapes, i.e., those that do not use MARS preprocessing. I haven't looked at the three video submissions so I don't know if their models are different or not.
